### PR TITLE
[Common] Optimize fused router forward/backward kernels

### PR DESCRIPTION
**Pull request by harryzhou2000**
## Summary

Optimizes the fused router CUDA kernels introduced in #2821 (`fused_topk_with_score_function` and `fused_score_for_moe_aux_loss`). Achieves significant bandwidth improvements for large expert counts and topk values while preserving identical performance for smaller configurations (e.g., E=256, topk=4).

**Key results (B300, float32, 8192 tokens):**
- Forward (E=2304, K=36, softmax): 673 → 964 GB/s (**+43%**)
- Backward (E=2304, K=36, softmax): 543 → 2766 GB/s (**+410%**)
- Forward (E=512, K=4): no regression (±0.3%)

### Forward kernels
- **Persistent grid with async double-buffered prefetch:** `RawAsyncLoader<T>` uses `cp.async` (sm_80+) for non-blocking global→shmem loads. Occupancy-aware grid sizing (`compute_persistent_grid`) keeps all SMs saturated across multiple rounds.
- **Packed 8-bit radix histogram:** Reduces radix topk register usage from 32 to 4 registers by packing 16 bucket counts into 4×u32 with 8-bit fields. Eliminates local memory spill at large E.
- **Compile-time score function dispatch:** `ScoreFunc` template parameter with `if constexpr` removes runtime branches from the hot loop.
- **Simple kernel path for small topk:** When `topk < NVTE_RADIX_TOPK_THRESHOLD` (default 8), dispatches to a lightweight kernel matching the original structure — no async loader, no persistent grid — avoiding scheduling overhead that dominates at small K.

### Backward kernels
- **Two-pass fused design:** Pass 1 accumulates warp-level sums via register reduction + `warp_allreduce_sum`. Pass 2 computes per-element gradients using scalar helpers. Eliminates the `comp_buf` shared memory buffer (saves `E × warps × 4` bytes per block).
- **Double-buffered async loading:** All backward inputs (grad, activation, mask) loaded through `RawAsyncLoader` with always-on double buffering.

### Infrastructure
- `async_loader.h`: `RawAsyncLoader<T>`, `compute_persistent_grid()`, `choose_num_buffers()`, vectorized global store/fill helpers.
- `NVTE_RADIX_TOPK_THRESHOLD` env var (default 8): configurable naive↔radix crossover.
- Templated `warp_reduce_on_shmem<T, ReduceFuncType>` eliminates function-pointer overhead.

### Hardening
- Host-side: `num_tokens * num_experts <= INT_MAX`, `topk ∈ [1, E]`, `topk % group_topk == 0`
- Device-side: `assert(data_size <= kMaxExpertsRadixTopk)` in radix path
- Correct `cudaDevAttrMaxSharedMemoryPerMultiprocessor` for buffer-count decision
- Fix: single-buffer prefetch clobber when shmem is too tight for double buffering

## Compatibility

- **No regression for small configs:** The simple forward kernel path is an exact replica of the original kernel structure, ensuring E=256/topk=4 (common in standard MoE) performs identically.
- **All existing tests pass:** 891/891 `test_fused_router.py` tests pass, 117 skipped (fp8/multi-node).
- **No API changes:** Same Python/C++ interface, same output semantics.
- **Tunable:** Set `NVTE_RADIX_TOPK_THRESHOLD=0` to force radix everywhere, or `=16` to use naive for topk<16.

## Performance (B300 SXM6, sm_103, float32, 8192 tokens)

Effective bandwidth (GB/s) is computed as the minimum bytes that must be transferred to/from global memory for one kernel invocation, divided by the measured wall time. For example, the topk forward kernel reads logits (`T×E×dtype`) and writes probs (`T×E×dtype`), routing_map (`T×E×1`), and intermediate_output (`T×E×4`). This metric captures how well the kernel utilizes memory bandwidth — higher is better, with the device peak around 8 TB/s on B300. Config format is `num_experts/topk`.

<details>
<summary>Full benchmark table (softmax)</summary>

| kernel   | pass  | config  | before   | after           |
| -------- | ----- | ------- | -------- | --------------- |
| topk     | fprop | 512/4   | 1779     | 1784 (+0.3%)    |
| topk     | fprop | 512/8   | 798      | 904 (+13%)      |
| topk     | fprop | 512/22  | 514      | 924 (+80%)      |
| topk     | fprop | 512/36  | 499      | 908 (+82%)      |
| topk     | fprop | 2304/4  | 1803     | 1802 (0%)       |
| topk     | fprop | 2304/8  | 660      | 993 (+51%)      |
| topk     | fprop | 2304/22 | 602      | 972 (+61%)      |
| topk     | fprop | 2304/36 | 673      | 964 (+43%)      |
| topk     | bprop | 512/22  | 3391     | 5362 (+58%)     |
| topk     | bprop | 2304/36 | 543      | 2766 (+410%)    |
| aux_loss | fprop | 512/22  | 519      | 896 (+73%)      |
| aux_loss | fprop | 2304/36 | 645      | 891 (+38%)      |
| aux_loss | bprop | 512/22  | 5289     | 6155 (+16%)     |
| aux_loss | bprop | 2304/36 | 2272     | 4201 (+85%)     |

</details>

<details>
<summary>Full benchmark table (sigmoid)</summary>

| kernel   | pass  | config  | before   | after           |
| -------- | ----- | ------- | -------- | --------------- |
| topk     | fprop | 512/4   | 1728     | 1736 (+0.5%)    |
| topk     | fprop | 512/22  | 470      | 891 (+90%)      |
| topk     | fprop | 2304/36 | 639      | 798 (+25%)      |
| topk     | bprop | 512/22  | 3169     | 4398 (+39%)     |
| topk     | bprop | 2304/36 | 533      | 2274 (+327%)    |
| aux_loss | fprop | 512/22  | 475      | 912 (+92%)      |
| aux_loss | fprop | 2304/36 | 598      | 867 (+45%)      |
| aux_loss | bprop | 2304/36 | 1965     | 2757 (+40%)     |

</details>

### CODE REVIEW

File: transformer_engine/common/fused_router/async_loader.h [hunk 1]
```diff
@@ -0,0 +1,256 @@
...
+      }
+      cp_async_commit();  // No-op on sm_70; matches wait() expectation on sm_80+.
+    }
+  }
+};
+
+}  // namespace fused_router
+}  // namespace transformer_engine
+
+#endif  // TRANSFORMER_ENGINE_FUSED_ROUTER_ASYNC_LOADER_H_
```
[the remark this answers is not in the document]
Comment (harryzhou2000) [reply transformer_engine/common/fused_router/async_loader.h:255]: Fixed in `3bab7cb1`. The scalar/unaligned fallback now only performs the synchronous copy and no longer calls `cp_async_commit()`, so it does not create an empty pipeline group on sm_80+.

File: transformer_engine/common/fused_router/fused_topk_with_score_function.cu [hunk 2]
```diff
@@ -244,36 +457,85 @@ void fused_topk_with_score_function_forward_kernel_launcher(
...
   } else {
-    NVTE_CHECK_CUDA(cudaFuncSetAttribute(
-        fused_topk_with_score_function_forward_kernel<DataType, BiasType, TopkFuncType::Radix>,
-        cudaFuncAttributeMaxDynamicSharedMemorySize, shared_memory_size));
-    fused_topk_with_score_function_forward_kernel<DataType, BiasType, TopkFuncType::Radix>
-        <<<grid_size, kThreadsPerBlock, shared_memory_size, stream>>>(
-            logits, num_tokens, num_experts, topk, use_pre_softmax, num_groups, group_topk,
-            scaling_factor, score_function, expert_bias, probs, routing_map, intermediate_output);
+    // Optimized path: async loader + persistent grid + radix topk.
+    NVTE_CHECK(num_experts <= kMaxExpertsRadixTopk,
```
Comment (denera) [inline transformer_engine/common/fused_router/fused_topk_with_score_function.cu]: Unless I'm misreading the code, this is going to hard-fail for `num_experts > 8160`.

@ptrendx Is this a restriction we can accept?

If not, I would recommend augmenting the algorithm selection (naive topk vs. radix-sort), and just simply fall back on the naive option whenever `num_experts > kMaxExpertsRadixTopk`. It may not be performant but it's probably better than a hard failure.

File: transformer_engine/common/fused_router/fused_score_for_moe_aux_loss.cu [hunk 3]
```diff
@@ -127,49 +246,91 @@ __global__ void fused_score_for_moe_aux_loss_forward_kernel(const DataType *logi
...
   } else {
-    NVTE_CHECK_CUDA(cudaFuncSetAttribute(
-        fused_score_for_moe_aux_loss_forward_kernel<DataType, TopkFuncType::Radix>,
-        cudaFuncAttributeMaxDynamicSharedMemorySize, shared_memory_size));
-    fused_score_for_moe_aux_loss_forward_kernel<DataType, TopkFuncType::Radix>
-        <<<grid_size, kThreadsPerBlock, shared_memory_size, stream>>>(
-            logits, num_tokens, num_experts, topk, score_function, scores, routing_map,
-            intermediate_output);
+    // Optimized path: async loader + persistent grid + radix topk.
+    NVTE_CHECK(num_experts <= kMaxExpertsRadixTopk,
```
Comment (denera) [inline transformer_engine/common/fused_router/fused_score_for_moe_aux_loss.cu]: Same issue as `fused_topk_with_score_function`. We should probably fall back on the naive path here instead of hard-failing when `num_experts > 8160`.

**Review by denera — CHANGES_REQUESTED** [review]

File: transformer_engine/common/fused_router/fused_topk_with_score_function.cu (hunk 2 above)
Comment (harryzhou2000) [reply transformer_engine/common/fused_router/fused_topk_with_score_function.cu]: Fixed in `2f17b7fe`. The launcher now selects radix only when `topk >= NVTE_RADIX_TOPK_THRESHOLD` and `num_experts <= kMaxExpertsRadixTopk`; otherwise it falls back to the existing naive path instead of hard-failing.

File: transformer_engine/common/fused_router/fused_score_for_moe_aux_loss.cu (hunk 3 above)
Comment (harryzhou2000) [reply transformer_engine/common/fused_router/fused_score_for_moe_aux_loss.cu]: Fixed in `2f17b7fe` with the same radix/naive dispatch condition as `fused_topk_with_score_function`, so `num_experts > kMaxExpertsRadixTopk` now uses the naive path.

Comment (harryzhou2000) [reply transformer_engine/common/fused_router/fused_score_for_moe_aux_loss.cu]: @denera I think the changes now seem complete and large `num_experts` now fall back on the naive path safely. Please take a look at [2f17b7f](https://github.com/NVIDIA/TransformerEngine/pull/3012/commits/2f17b7fee25be6ed5f58a6adcb29ed165e02331a).

**Review by denera — DISMISSED** [review]
LGMT. I'll fire off the CI once you resolve merge conflicts with main and then we can merge as soon as it clears through. Thanks!

**Review by denera — APPROVED** [review]
LGTM, CI errors are unrelated to this branch, merging.
